pkg/modules/modules.go

```diff
@@ -0,0 +1,162 @@
+package modules
+
+import (
+	"context"
+	"errors"
+
+	"github.com/grafana/dskit/modules"
+	"github.com/grafana/dskit/services"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+// List of available targets.
+const (
+	BackgroundServices string = "background"
+	All                string = "all"
+)
+
+type Engine interface {
+	Init(context.Context) error
+	Run(context.Context) error
+	Shutdown(context.Context) error
+}
+
+type Manager interface {
+	RegisterModule(name string, initFn func() (services.Service, error), deps ...string)
+	RegisterInvisibleModule(name string, initFn func() (services.Service, error), deps ...string)
+}
+
+var _ Engine = (*service)(nil)
+var _ Manager = (*service)(nil)
+
+// service manages the registration and lifecycle of modules.
+type service struct {
+	cfg           *setting.Cfg
+	log           log.Logger
+	targets       []string
+	dependencyMap map[string][]string
+
+	ModuleManager  *modules.Manager
+	ServiceManager *services.Manager
+	ServiceMap     map[string]services.Service
+}
+
+func ProvideService(cfg *setting.Cfg) *service {
+	logger := log.New("modules")
+
+	return &service{
+		cfg:           cfg,
+		log:           logger,
+		targets:       cfg.Target,
+		dependencyMap: map[string][]string{},
+
+		ModuleManager: modules.NewManager(logger),
+		ServiceMap:    map[string]services.Service{},
+	}
+}
+
+// Init initializes all registered modules.
+func (m *service) Init(_ context.Context) error {
+	var err error
+
+	// module registration
+	m.RegisterModule(All, nil)
+
+	// module dependencies
+	m.dependencyMap = map[string][]string{
+		BackgroundServices: {},
```

Comment:
🤔 does this manual mapping need to happen in `ProvideService` in order to support registering with the `RegisterModule` and `RegisterInvisibleModule` methods?

Reply:
Do you mean does this dependency mapping need to be created before the background service registers itself? Not currently no - but I realize I was mixing patterns here. This shouldn't be overwriting the map, but just defining the static ones (IE just `All` for now), allowing the `Register*` methods to alter `dependencyMap` where necessary